lib/common/output_html.c

```diff
@@ -0,0 +1,318 @@
+     * "stylesheet-link" option, and this should obviously be a link to a
+     * stylesheet.  The second can override the first.  At least one should be
+     * given.
+     */
+    xmlNewTextChild(head_node, NULL, (pcmkXmlStr) "style", (pcmkXmlStr) stylesheet_default);
+
+    if (stylesheet_link != NULL) {
+        htmlNodePtr link_node = xmlNewChild(head_node, NULL, (pcmkXmlStr) "link", NULL);
+        xmlSetProp(link_node, (pcmkXmlStr) "rel", (pcmkXmlStr) "stylesheet");
+        xmlSetProp(link_node, (pcmkXmlStr) "href", (pcmkXmlStr) stylesheet_link);
+    }
+
+    xmlAddPrevSibling(priv->root->children, head_node);
+
+    htmlDocDump(out->dest, priv->root->doc);
+}
+
+static void
+html_reset(pcmk__output_t *out) {
+    private_data_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+
+    htmlDocDump(out->dest, priv->root->doc);
+
+    html_free_priv(out);
+    html_init(out);
+}
+
+static void
+html_subprocess_output(pcmk__output_t *out, int exit_status,
+                      const char *proc_stdout, const char *proc_stderr) {
+    char *rc_buf = NULL;
+    private_data_t *priv = out->priv;
+    CRM_ASSERT(priv != NULL);
+
+    rc_buf = crm_strdup_printf("Return code: %d", exit_status);
+
+    xmlNewTextChild(g_queue_peek_tail(priv->parent_q), NULL,
+                    (pcmkXmlStr) "h2", (pcmkXmlStr) "Command Output");
+
+    pcmk__xml_add_node(out, pcmk__html_mk_div(rc_buf, NULL, NULL));
+
+    if (proc_stdout != NULL) {
+        pcmk__xml_add_node(out, pcmk__html_mk_div("Stdout", NULL, NULL));
+        pcmk__xml_add_node(out, pcmk__html_mk_div(proc_stdout, "output", NULL));
+    }
+    if (proc_stderr != NULL) {
+        pcmk__xml_add_node(out, pcmk__html_mk_div("Stderr", NULL, NULL));
+        pcmk__xml_add_node(out, pcmk__html_mk_div(proc_stderr, "output", NULL));
+    }
+
+    free(rc_buf);
+}
+
+G_GNUC_PRINTF(2, 3)
+static void
+html_info(pcmk__output_t *out, const char *format, ...) {
+    /* This function intentially left blank */
+}
+
+static void
+html_output_xml(pcmk__output_t *out, const char *name, const char *buf) {
+    htmlNodePtr node = NULL;
+    private_data_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+
+    node = xmlNewTextChild(g_queue_peek_tail(priv->parent_q), NULL,
+                           (pcmkXmlStr) "script", (pcmkXmlStr) buf);
+    xmlSetProp(node, (pcmkXmlStr) "id", (pcmkXmlStr) name);
+    xmlSetProp(node, (pcmkXmlStr) "type", (pcmkXmlStr) "application/xml");
+}
+
+static void
+html_begin_list(pcmk__output_t *out, const char *name,
+               const char *singular_noun, const char *plural_noun) {
+    htmlNodePtr list_node = NULL;
+    private_data_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+
+    if (name != NULL) {
+        xmlNewChild(g_queue_peek_tail(priv->parent_q), NULL, (pcmkXmlStr) "h2",
+                    (pcmkXmlStr) name);
+    }
+
+    list_node = create_xml_node(g_queue_peek_tail(priv->parent_q), "ul");
```

Comment:
I have been pretty sloppy with when I use one vs. another.  There's also xmlNewTextChild, too.  I should probably standardize what I am doing now before dropping the crm_mon formatted output patches on everyone because those make a whole lot more XML nodes.

Reply:
What do you think about adding a new function like this:
```xmlNode *create_xml_text_node(xmlnode *parent, const char *name, const char *content)```

It would act the same as create_xml_node, but also add the given content as the text content of the node.  It could be implemented in terms of create_xml_node.  I think I could then replace all uses of xmlAddChild, xmlNewChild, and xmlNewTextChild with one of the two create_xml_ functions.  I could add a patch to this PR to add the new function and switch all uses, or it could be a different PR.

Reply:
That works for me. Doing it in this PR makes sense.